### ISSUE TITLE
Swift REPL

### ISSUE THREAD

**Issue by jmjwozniak** [swift-lang/swift-t#62]
REPL model for Swift/T.

**Comment by basheersubei** [comment]
I'm starting to tinker with the Turbine API and calling std input from within a Swift leaf function [here](https://github.com/basheersubei/swift-t-testing/commit/710394a2bdc79492c63812683a18f3ecd38c05dc).

@jmjwozniak I was wondering, is there a way to get variable ids or probe for data members using the Turbine API? Currently, my REPL loop is sitting in a Swift leaf function, and I don't have references to the variables. Should I write functionality that probes and gets all ADLB data member ids?

**Comment by basheersubei** [comment]
since all the generated ids of ADLB data members seem to be consecutive (1, 2, 3, 4... etc), can't I just call a `adlb::retrieve` on numbers starting from zero until we run out?

it seems like a terribly hacky way of doing it, but I can't think of anything else (perhaps due to my lack of knowledge of adlb internals)

**Comment by timarmstrong** [comment]
I'll jump in if that's not a problem.

If there are multiple ADLB servers they are not assigned consecutively like
that, but even with a single server you're going to run into more problems
soon.

Looking at the code, the intent is that you should be able to refer to the
global variable a from within the Tcl function without explicitly passing a
into repl_tcl(), right?

You will have trouble achieving this without modification to the compiler.
The compiler has no idea that you intended to implicitly pass a into
repl_tcl(), so it will feel free to optimise a out.  Looking at the
compiled code you checked in, I don't see any references to a, so that's
what happened.

The REPL can presumably keep track of any variables it created itself so
it's mainly a problem of getting predeclared variables into the REPL.

I think the options are:
a) explicit pass in any variables you want the REPL to have access to.
This seems easiest
b) modify the compiler so that all global variables are implicitly passed
into certain functions like repl_tcl().  I can help spec out the changes if
that would be useful, but it involves getting into some of the guts of the
compiler.
c) disable the dead code elimination (and possibly other) optimisations in
REPL mode and add functionality to ADLB to enumerate variables.  I'm not
sure if there's enough debug information attached to ADLB variables to
actually determine whether a random data item you pulled out of ADLB is a
Swift global, though...

**Comment by basheersubei** [comment]
relevant ADLB functions (these can all be called from within Tcl inside the turbine namespace):
* retrieve (gets a data member given an id)
* exists (checks whether data given an id exists)
* closed (checks whether data given an id is closed [is assigned a value?])
* locate (returns which server data given an id is located)
* servers (returns number of servers)

**Comment by basheersubei** [comment]
oh hey @timarmstrong, thanks for your advice!

I was thinking to completely ignore optimizations for now, perhaps just in the initial stages. But now that you mention it, option a) sounds like a very good idea! If the REPL were incorporated as part of the Turbine engine (or maybe even externally), it could keep track of the all the variables (since they all happen within the REPL, whether if the user interactively types them in or tells it to run a script). I guess I need to sit down and flesh this out more conceptually.

Much appreciated! :)

**Comment by basheersubei** [comment]
I have no idea why I was thinking that the REPL needs to sort of "attach" itself to an existing Swift session (or predeclared variables) and map out all the variables... There is no need for this; the user will run the REPL environment and then they can run Swift code interactively. (still not sure how exactly but I'll think about it more)

**Comment by jmjwozniak** [comment]
I would try the "STC object file" subproject first- I think that will be a better learning curve.

**Comment by basheersubei** [comment]
@jmjwozniak I'm not sure what you mean by STC object file. Is there a webpage where it's discussed in detail?

**Comment by basheersubei** [comment]
Meanwhile, I've been working on a very basic version of the REPL. Current implemented features:
* User can type in swift code, and then type %END to signify that he's finished typing code.
* The REPL (embedded in the turbine engine as a leaf function) writes the swift code to a file, calls STC on it, and reads in the .tic output.
* (Partially finished) parsing through the .tic output to remove boilerplate code (like calls to `turbine::init` and `turbine::finalize`). I'm using regex for this and it seems to be working pretty well.

Remaining feature:
* once the .tic is processed, the REPL will just execute that tic code (with a simple Tcl `eval`).

**Comment by basheersubei** [comment]
![anim](https://cloud.githubusercontent.com/assets/6508762/8685086/ec989a56-2a41-11e5-9778-591f769586db.gif)
 
I have a basic version that works with simple scripts. So far I can run the scripts within Turbine (still can't tap into global variable states). It crashes when I attempt to spawn too many tasks (around 65).

Am I on the right track? Or is this too hacky? (I'm basically just `eval`ing .tic code inside a Swift leaf function).

**Comment by basheersubei** [comment]
(assuming unoptimized STC)
so I noticed that the .tic output gives global variables an id in turbine. If I run through the .tic output (from whatever Swift user input compiled thru STC), I can get those ids and relate them to variable names declared in my REPL. Then, if the user asks to see the state of variable x, I can just run `adlb::retrieve <id>` with that id.

**Comment by timarmstrong** [comment]
Looks like awesome progress.  As you've probably noticed, it wasn't really
designed with a REPL in mind so there will definitely be some improvisation
- hopefully we can add hooks to the underlying system as we go to make it
better/more streamlined.

As you noticed, there is a concept of global variables in Turbine/ADLB -
they're generally initialized when the script starts up and assigned
negative identifiers (which is really just an implementation detail).

The global names aren't actually stored anywhere at runtime but it could be
arranged.  If you want to get a hold of a list of those globals then, say,
if you implemented a Tcl function in turbine/lib/turbine.tcl that took a
list of ADLB IDs and a list of variable names I could modify the code
generator so that it called that function.  It's pretty easy to get it to
emit more Tcl statements since it already has the names and IDs.  The code
generator isn't anything super fancy so you could probably figure it out
but it is just a bunch of ugly Java code so not sure that it's a good use
of your time.

Would something like that help you make progress?

- Tim

**Comment by basheersubei** [comment]
That sounds like a much better idea than what I was thinking!

ok, just to make sure I understand you correctly:
1- user enters swift code, something like `int x = 1;`
2- STC compiles it and then generates an additional call to a turbine/ADLB function that stores variable names with that given id.
3- the generated turbine code is called, including this new function (which stores the name along with the id, creating some sort of association).
4- Now my REPL code (later on, when user requests variable status) can fetch ADLB data by name (by using the name to fetch the id). Some function call will have to be written to retrieve id by name.

^ Is that right?

But where exactly in Turbine/ADLB would the name be stored? I can see there's an `adlb_datum_id` defined for all ADLB data objects, and I know data is stored on each ADLB server separately. Maybe we can add a name member in the `ADLB_create_spec` struct. Do I have to worry about any of the ADLB data store magic (I pretty much have no idea how any of that works)?

**Comment by basheersubei** [comment]
also, why not extend this ability to all variables, not just global ones? Or are the ids assigned to each variable (across different ADLB servers) not unique?

**Comment by timarmstrong** [comment]
I was just thinking of a way, where, if you evaluate a .tic file from
within your Tcl interpreter, you can get a hold of the name->id mapping for
global variables.  A  name->id map could be stored as a Tcl dictionary or
similar for the purposes of the REPL.  I think the REPL can manage that map
ok without any involvement from the underlying systems so long as it
actually has a way to obtain the name/id mappings.

I don't see why the REPL would need to know the IDs of local variables in
Swift functions - I don't think REPLs usually let you step through
functions, do they?

There's actually a mechanism to attach debug information to variables -
each variable has an integer symbol attached that can be used to look up
debug info with the *dsym functions in adlb.h:
https://github.com/swift-lang/swift-t/blob/master/lb/code/src/adlb.h.  We
*don't* actually have a way to get the symbol from the client (we probably
ideally would) but I think this is tangential to the REPL problem.

Oh, and variable IDs are unique, but each local variable name could map to
multiple IDs at each point in time.

- Tim

**Comment by basheersubei** [comment]
ok, I think I understand how to do it now. Before `eval`ing the .tic file, I can just go through it, get the global variable names (all found in the `adlb::create_globals` call), and then add a call to a function (at the end of the .tic code) that creates a dictionary mapping names->ids. I have to add this function call because the ids are only available during runtime (I can't get them by running through the .tic code). There's no need to store that dictionary on the turbine engines, like you said.

You're right about the local variables. If anything, it's an advanced feature that I won't worry about now.


After talking to Justin and Mike, I think I'll disable optimizations for now (easier to make progress with the REPL), but I'll also start thinking about the "STC object file" idea (to allow STC to compile fragments of Swift code like function definitions into these object-like files, only to later link it to other object files, like a main Swift script that calls those functions). Working on this should make it clearer how I'll handle optimizations for the REPL, I think.

**Comment by timarmstrong** [comment]
Are you proposing to grep the .tic file?  I'm not sure that's a good idea -
it'll be brittle.  I was suggesting to add a proper Tcl interface that
would let you get the global variables.

Maybe you can hardcode the variable list for your test program just to get
started and I'll add in that support today or tomorrow?

**Comment by basheersubei** [comment]
How would you achieve that, though? How would you "get a hold of the name->id mapping for
global variables" from the .tic file? I'm not sure how I'd make a proper Tcl interface. (I agree that my idea makes it brittle). Sorry if I'm not grasping this quickly enough.

**Comment by timarmstrong** [comment]
I'm going to modify the output of STC so that instead of
adlb::create_globals it calls a function turbine::declare_globals .
 turbine::declare_globals will call adlb::create_globals then also update a
data structure that stores the mapping for later use..

I'll add a function that you can call from your Tcl code to get this info.

**Comment by basheersubei** [comment]
@timarmstrong I'm having a bit of trouble getting the `globals_map` working.

I made a repl version of `multicreate` called [`turbine::multicreate_repl`](https://github.com/basheersubei/swift-t/commit/e7e90034703fa0d77db14fc6829ba991cc586c66). I can't get it to create the globals_map on all the workers (it works for rank 0, which creates the globals_map).

If you run the test (which is just like turbine-globals.tcl, except uses `multicreate_repl` instead of `declare_globals`, you can see it failing whenever you have more than 1 worker. For some reason, all the other workers return an empty list whenever I call `turbine::get_globals_map`.

Could it perhaps be a namespace thing, particularly [where the `globals_var` dict is being appended to?](https://github.com/basheersubei/swift-t/commit/e7e90034703fa0d77db14fc6829ba991cc586c66#diff-8871ae3ad36537e01bebb798cd5e04acR651) I pretty much copied `declare_globals` but added an `if{rank==0}` around the dict creation.

**Comment by timarmstrong** [comment]
I haven't had a chance to properly look at everything you sent over - will
respond in more detail tomorrow.  Tcl "variable" variables are scoped to a
namespace so as long as it's in the turbine namespace it should be ok.

- Tim

**Comment by timarmstrong** [comment]
How are the allocated IDs communicated from rank 0 to the other ranks? I think you need to create tasks on rank 0 to send that info to the other ranks. 

I'm not sure if you know this, but you can pretty easily serialize Tcl procedure calls by putting the proc name and arguments in a list, like:
[ list the_function $arg1 [ list $list_element1 $list_element2 ] ]

**Comment by timarmstrong** [comment]
How are the allocated IDs communicated from rank 0 to the other ranks? I think you need to create tasks on rank 0 to send that info to the other ranks. 

I'm not sure if you know this, but you can pretty easily serialize Tcl procedure calls by putting the proc name and arguments in a list, like:
[ list the_function $arg1 [ list $list_element1 $list_element2 ] ]

**Comment by timarmstrong** [comment]
Sorry about that, accidentally closed it.

**Comment by basheersubei** [comment]
you're absolutely right, I can't believe I missed that. Rank 0 should put tasks to define the globals on the other ranks before putting tasks to run everything.

Initial testing appears to fix the problem. Now for a proper fix. Hopefully I'll have it working by the end of the day.

**Comment by timarmstrong** [comment]
Nice work :). It's a really weird programming model with everything being asynchronous and running in unpredictable locations.

**Comment by basheersubei** [comment]
@timarmstrong 
a quick question:
what exactly does the parallelism flag for `adlb::put` mean?

**Comment by jmjwozniak** [comment]
Number of workers per task- fix this at 1 for now.

**Comment by timarmstrong** [comment]
Yep, parallelism = 1 is the right value unless you're doing something very specific. parallelism > 1 the task has to be executed simultaneously by multiple workers.  ADLB gathers n workers together, sets up a shared MPI communicator, then sends the task to all the workers, so the workers can communicate with each other. We published a paper about it - http://swift-lang.org/papers/pdfs/Swift_MPI_2013.pdf

**Comment by basheersubei** [comment]
now the REPL can run a swift script and correctly put tasks to multiple workers (keeping the proc and global definitions across all workers). I can also look at the status/values of globals from the REPL using the `%ls` command.

However, it seems the next step is to get subsequent scripts to work together properly. I'm assuming the REPL will (ultimately, in the final implementation) process each line of input (or multi-line for functions and loops) separately. i.e. it will have to compile then load tic for each line of input. This introduces some problems.

Some use cases with possible solutions:

1. user inputs two scripts, with the second script relying on definitions from first script (global variable, function, or even import statement). 
    * STC will complain about missing definitions. **Solution:** we could implicitly place the definitions in the subsequent scripts.
    * proc name collisions in the tic code (STC might use same names). **Solution:** make STC generate unique proc names when `-c` flag is used. Global names will not be an issue as they are unique by definition.

2. user wants to redefine variables/functions declared in previous scripts.
    * this conflicts with above solution of implicitly placing definitions (STC will freak out when given multiple definitions of globals/functions).

Next step: look at more formal definition of use cases, and also at how different REPL environments deal with this. Specifically, IPython and JuPyter.

**Comment by basheersubei** [comment]
@timarmstrong @jmjwozniak  I'm trying to get STC to **not** generate the location global variables `HARD`, `SOFT`, `RANK`, and `NODE`. They're usually removed during optimization, but I'm using `-O0` and I'd rather not generate them in STC than having to remove them later post-STC.

Any idea where they're added to the list of global variables?

**Comment by basheersubei** [comment]
also update: the REPL currently **prepends** extern statements for all global variables for every Swift script the user enters. This means you can reuse variables you've defined in earlier scripts.

**Comment by timarmstrong** [comment]
They're in the builtins.swift header in Turbine.  STC automatically loads that before any other files. See: https://github.com/swift-lang/swift-t/blob/master/stc/code/src/exm/stc/frontend/ASTWalker.java#L169

The challenge is that you probably need to process the definitions so that you know variable/function types defined in builtins.swift, but you don't want to actually compile the top-level code. A possible fix is to skip the compilation of top-level code from the builtins module when -c is enabled (or something along those lines, since I don't know all of the details of -c). I think this is a specific instance of a general problem that you don't necessary want to recompile all imported modules every time STC is invoked.

**Comment by basheersubei** [comment]
On second thought (after about an hour of digging through STC stuff), I decided it would be much easier to just do it post-STC in the repl code, because it's only these four globals and I won't be needing a general solution.

**Comment by basheersubei** [comment]
Update: Was much easier to do in Tcl REPL.

Now, to create an IPython/Jupyter Swift kernel, so that the REPL can be accessed through IPython clients (shell sessions, notebooks, etc.)

**Comment by timarmstrong** [comment]
That would be awesome.

**Comment by basheersubei** [comment]
Initial attempts to control the turbine repl process from Python (jupyter kernel) using pipes were unsuccessful. Apparently, it's not a good idea to use pipes for two-way communication between long-running processes. I'm switching over to using sockets (can't think of anything else).

**Comment by basheersubei** [comment]
Much easier now using sockets. Finished the python side (jupyter kernel), which now acts as a socket server. Let's hope sockets in Tcl are easy... [uneasiness sets in] xD
